libcontainer/cgroups/fs/blkio.go

```diff
@@ -166,13 +166,83 @@ func getBlkioStat(path string) ([]cgroups.BlkioStatEntry, error) {
 }
 
 func (s *BlkioGroup) GetStats(path string, stats *cgroups.Stats) error {
-	// Try to read CFQ stats available on all CFQ enabled kernels first
+	// Try to read BFQ stats with debug files enabled first
+	if blkioStats, err := getBlkioStat(filepath.Join(path, "blkio.bfq.io_service_time_recursive")); err == nil && blkioStats != nil {
+		return getBFQDebugStats(path, stats)
+	}
+	// Try to read BFQ stats with debug files disabled
+	if blkioStats, err := getBlkioStat(filepath.Join(path, "blkio.bfq.io_serviced_recursive")); err == nil && blkioStats != nil {
+		return getBFQStats(path, stats)
+	}
+	// Try to read CFQ stats available on all CFQ enabled kernels
```

Comment:
It seems you're reading the same file twice, and the first read is just to check that the file is available.

It might be better to just call the appropriate function and fall back to another one if it fails. Or, just check that the file exist, instead of reading it.

Reply:
> It might be better to just call the appropriate function and fall back to another one if it fails.

If it fails on any file after first one, it means that it uses that scheduling method and just got corrupted file, falling back to another method in that case would just waste time.

>Or, just check that the file exist, instead of reading it.

File may exist, but be empty or contain only `Total: 0` (I'm sure this happens at least with CFQ present but not enabled)

Reply:
File is read only once